File: deliverables/official-docs/src/official-docs.js
/**
 * Official documentation registry — single source of truth.
 * ZyntroAI standard: v1.2.0-origin
 *
 * Every URL in OFFICIAL_DOCS was verified with a live HTTP request and the
 * verification date is recorded in VERIFIED_ON. URLs the original brief cited
 * that do NOT resolve are kept in CORRECTED_LINKS together with their
 * replacement, so the audit trail survives the correction.
 *
 * Consumers:
 *   - frontend:  Company.jsx imports OFFICIAL_DOCS / analyze / getSourceDoc
 *   - backend:   official_docs.py reads the generated official-docs.json
 *   - CI:        verify_links.py re-checks every URL
 */

export const REGISTRY_VERSION = "1.2.0-origin";
export const VERIFIED_ON = "2026-09-14";

export const LINK_STATUS = {
  LIVE: "live",
  REPLACED: "replaced",
  REMOVED: "removed",
};

export const IMAGE_EXTENSIONS = ["png", "jpg", "jpeg", "webp", "gif", "svg", "avif"];

export const OFFICIAL_DOCS = {
  fastapi: {
    name: "FastAPI Official",
    label: "Core Framework",
    source: "FastAPI",
    links: [
      { id: "docs", label: "Official Docs", url: "https://fastapi.tiangolo.com/" },
      { id: "repo", label: "GitHub Repo", url: "https://github.com/tiangolo/fastapi" },
      { id: "advanced", label: "Advanced Guide", url: "https://fastapi.tiangolo.com/advanced/" },
    ],
  },
  python: {
    name: "Python Docs",
    label: "Language Reference",
    source: "Python",
    links: [{ id: "docs", label: "Python 3 Docs", url: "https://docs.python.org/3/" }],
  },
  fig: {
    name: "FIG Platform",
    label: "Asset Source",
    source: "FIG",
    links: [
      { id: "site", label: "Official Site", url: "https://hellofig.app/" },
      { id: "share", label: "Share / Assets", url: "https://share.hellofig.app/" },
    ],
  },
  dola: {
    name: "Dola AI Vision",
    label: "Image Intelligence",
    source: "Dola",
    links: [
      { id: "site", label: "Dola AI", url: "https://dola.ai/" },
      {
        id: "image-analysis",
        label: "Image Analysis Docs",
        url: "https://dola.ai/docs/features/image-analysis",
      },
      { id: "api", label: "API Reference", url: "https://dola.ai/api-reference" },
    ],
  },
  zyntroai: {
    name: "ZyntroAI Origin Standard",
    label: "Internal Standard",
    source: "ZyntroAI",
    links: [
      {
        id: "origin",
        label: "Origin Branch",
        url: "https://github.com/ZyntroAI/fastapi-python-boilerplate/tree/Origin",
      },
      {
        id: "pr-template",
        label: "PR Template",
        url: "https://github.com/ZyntroAI/fastapi-python-boilerplate/blob/Origin/.github/PULL_REQUEST_TEMPLATE.md",
      },
    ],
  },
};

/**
 * Links from the original brief that were dropped or rewritten, and why.
 * `replacement: null` means nothing usable existed — do not invent one.
 */
export const CORRECTED_LINKS = [
  {
    cited: "https://fastapi.tiangolo.com/advanced/architecture/",
    reason: "404 — that path does not exist",
    replacement: "https://fastapi.tiangolo.com/advanced/",
  },
  {
    cited: "https://docs.hellofig.ai/",
    reason: "Host does not resolve (DNS failure)",
    replacement: "https://hellofig.app/",
  },
  {
    cited: "https://share.hellofig.app/help",
    reason: "404 — /help is not a help centre; the host serves shared conversations",
    replacement: "https://share.hellofig.app/",
  },
  {
    cited: "https://hellofig.app/changelog",
    reason: "404 — no public changelog at this path",
    replacement: null,
  },
  {
    cited: "https://hellofig.ai/",
    reason: "Different domain from the official hellofig.app platform site",
    replacement: "https://hellofig.app/",
  },
  {
    cited:
      "https://github.com/ZyntroAI/fastapi-python-boilerplate/blob/Origin/.github/CHECKLIST_BILLING.md",
    reason: "404 — file does not exist on Origin or main",
    replacement: null,
  },
];

/** source name (as detected from an image URL) -> OFFICIAL_DOCS key */
export const SOURCE_KEYS = {
  FastAPI: "fastapi",
  Python: "python",
  FIG: "fig",
  Dola: "dola",
  ZyntroAI: "zyntroai",
};

/** Accept only http/https — blocks javascript: and data: hrefs. */
export function isSafeUrl(url) {
  try {
    const parsed = new URL(url);
    return parsed.protocol === "http:" || parsed.protocol === "https:";
  } catch {
    return false;
  }
}

export function getPrimaryUrl(key) {
  const entry = OFFICIAL_DOCS[key];
  return entry && entry.links.length > 0 ? entry.links[0].url : null;
}

export function getSourceDoc(source) {
  const key = SOURCE_KEYS[source];
  return key ? OFFICIAL_DOCS[key] : null;
}

/**
 * Map an asset URL to the provider it came from.
 * Falls back to string sniffing when the value is not a parseable URL.
 */
export function detectSource(src) {
  const raw = String(src == null ? "" : src);
  let host = "";
  let path = "";
  try {
    const parsed = new URL(raw);
    host = parsed.hostname.toLowerCase();
    path = parsed.pathname.toLowerCase();
  } catch {
    const lower = raw.toLowerCase();
    if (lower.includes("dola.ai")) return "Dola";
    if (lower.includes("hellofig")) return "FIG";
    if (lower.includes("fastapi")) return "FastAPI";
    if (lower.includes("python.org")) return "Python";
    if (lower.includes("zyntroai")) return "ZyntroAI";
    return "Generic";
  }

  if (host === "dola.ai" || host.endsWith(".dola.ai")) return "Dola";
  if (host === "hellofig.app" || host === "hellofig.ai" || host.endsWith(".hellofig.app")) {
    return "FIG";
  }
  if (host === "docs.python.org" || host.endsWith(".python.org")) return "Python";
  if (host === "fastapi.tiangolo.com") return "FastAPI";
  if (host === "github.com") {
    if (path.startsWith("/tiangolo/fastapi")) return "FastAPI";
    if (path.startsWith("/zyntroai/")) return "ZyntroAI";
  }
  return "Generic";
}

/**
 * Describe an image asset: its provider, the official doc it should link to,
 * and structural facts. Pure and synchronous — safe to call during render.
 */
export function analyze(src, meta) {
  const extra = meta || {};
  const raw = String(src == null ? "" : src);
  const source = detectSource(raw);
  const doc = getSourceDoc(source);

  let host = "";
  let pathname = "";
  let ext = "";
  try {
    const parsed = new URL(raw);
    host = parsed.hostname;
    pathname = parsed.pathname;
    const match = /\.([a-z0-9]+)$/i.exec(pathname);
    if (match) ext = match[1].toLowerCase();
  } catch {
    const match = /\.([a-z0-9]+)(?:[?#].*)?$/i.exec(raw);
    if (match) ext = match[1].toLowerCase();
    pathname = raw;
  }

  return {
    src: raw,
    source,
    docRef: doc ? getPrimaryUrl(SOURCE_KEYS[source]) : null,
    docName: doc ? doc.name : null,
    host,
    pathname,
    ext,
    isImage: IMAGE_EXTENSIONS.includes(ext),
    isExternal: /^https?:\/\//i.test(raw),
    safe: isSafeUrl(raw),
    alt: extra.alt || "",
  };
}

/** Image skill descriptor — the documented surface other modules import. */
export const IMAGE_SKILL = {
  name: "ZyntroAI Vision Engine",
  version: REGISTRY_VERSION,
  sources: Object.keys(OFFICIAL_DOCS),
  docs: OFFICIAL_DOCS,
  verifiedOn: VERIFIED_ON,
  analyze,
  detectSource,
  getSourceDoc,
  getPrimaryUrl,
  isSafeUrl,
};

/** Machine-readable mirror handed to official_docs.py and CI. */
export function toRegistry() {
  return {
    version: REGISTRY_VERSION,
    verifiedOn: VERIFIED_ON,
    docs: OFFICIAL_DOCS,
    sourceKeys: SOURCE_KEYS,
    corrections: CORRECTED_LINKS,
    imageExtensions: IMAGE_EXTENSIONS,
  };
}

export default OFFICIAL_DOCS;

File: deliverables/official-docs/src/official_docs.py
"""Official documentation registry — Python twin of src/official-docs.js.

Same detection and safety logic as the JavaScript/TypeScript versions, adapted
to the backend's I/O model: the registry is loaded from the generated
official-docs.json (exported by scripts/export_registry.mjs) rather than
imported as a module graph, and the "which source?" question is answered from
ASGI request headers instead of DOM attributes.

Standard library only.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Iterable, Mapping
from urllib.parse import urlsplit

DEFAULT_REGISTRY = Path(__file__).with_name("official-docs.json")

IMAGE_EXTENSIONS = frozenset({"png", "jpg", "jpeg", "webp", "gif", "svg", "avif"})

SOURCE_KEYS: dict[str, str] = {
    "FastAPI": "fastapi",
    "Python": "python",
    "FIG": "fig",
    "Dola": "dola",
    "ZyntroAI": "zyntroai",
}

__all__ = [
    "DocRegistry",
    "detect_source",
    "source_from_headers",
    "is_safe_url",
    "analyze",
    "load_registry",
]


def is_safe_url(url: str | None) -> bool:
    """Accept only http/https. Mirrors ``isSafeUrl()``."""
    if not url:
        return False
    try:
        parts = urlsplit(str(url))
    except ValueError:
        return False
    return parts.scheme in ("http", "https") and bool(parts.netloc)


def _host_and_path(raw: str) -> tuple[str, str]:
    try:
        parts = urlsplit(raw)
    except ValueError:
        return "", ""
    return parts.hostname or "", parts.path.lower()


def detect_source(src: str | None) -> str:
    """Map an asset URL to the provider it came from.

    Falls back to string sniffing when the value is not a parseable URL —
    identical behaviour to the JS implementation.
    """
    raw = "" if src is None else str(src)
    if not raw:
        return "Generic"

    host, path = _host_and_path(raw)
    if not host:
        lower = raw.lower()
        if "dola.ai" in lower:
            return "Dola"
        if "hellofig" in lower:
            return "FIG"
        if "fastapi" in lower:
            return "FastAPI"
        if "python.org" in lower:
            return "Python"
        if "zyntroai" in lower:
            return "ZyntroAI"
        return "Generic"

    if host == "dola.ai" or host.endswith(".dola.ai"):
        return "Dola"
    if host in ("hellofig.app", "hellofig.ai") or host.endswith(".hellofig.app"):
        return "FIG"
    if host == "docs.python.org" or host.endswith(".python.org"):
        return "Python"
    if host == "fastapi.tiangolo.com":
        return "FastAPI"
    if host == "github.com":
        if path.startswith("/tiangolo/fastapi"):
            return "FastAPI"
        if path.startswith("/zyntroai/"):
            return "ZyntroAI"
    return "Generic"


def source_from_headers(headers: Mapping[str, str]) -> str:
    """Backend adaptation of ``sourceFromElement()``.

    The browser reads a ``data-source`` attribute off the image element; an
    ASGI request has no DOM, so the equivalent declaration travels as the
    ``x-asset-source`` header and the fallback is ``Referer``.
    """
    lower = {str(k).lower(): str(v) for k, v in headers.items()}

    declared = lower.get("x-asset-source", "").strip()
    if declared:
        return declared

    referer = lower.get("referer") or lower.get("referrer") or ""
    if referer:
        return detect_source(referer)

    origin = lower.get("origin", "").strip()
    if origin:
        return detect_source(origin)

    return "Generic"


def analyze(src: str | None, *, alt: str = "", source: str | None = None) -> dict[str, Any]:
    """Describe an image asset: provider, official doc, structural facts.

    Mirrors the JavaScript ``analyze()`` result shape.
    """
    raw = "" if src is None else str(src)
    resolved = source or detect_source(raw)

    if is_safe_url(raw):
        parts = urlsplit(raw)
        host = parts.hostname or ""
        pathname = parts.path
    else:
        host, pathname = "", raw

    ext = ""
    for candidate in (pathname.lower(), raw.lower()):
        if "." in candidate:
            tail = candidate.rsplit(".", 1)[-1]
            tail = tail.split("#", 1)[0].split("?", 1)[0]
            if tail.isalnum():
                ext = tail
                break

    return {
        "src": raw,
        "source": resolved,
        "doc_key": SOURCE_KEYS.get(resolved),
        "host": host,
        "pathname": pathname,
        "ext": ext,
        "is_image": ext in IMAGE_EXTENSIONS,
        "is_external": raw.lower().startswith(("http://", "https://")),
        "safe": is_safe_url(raw),
        "alt": alt,
    }


class DocRegistry:
    """Loaded registry. Same knowledge as OFFICIAL_DOCS in the JS module."""

    def __init__(self, payload: Mapping[str, Any]) -> None:
        self.version: str = str(payload.get("version", ""))
        self.verified_on: str = str(payload.get("verifiedOn", ""))
        self.docs: dict[str, Any] = dict(payload.get("docs", {}))
        self.source_keys: dict[str, str] = dict(payload.get("sourceKeys", SOURCE_KEYS))
        self.corrections: list[dict[str, Any]] = list(payload.get("corrections", []))

    def __repr__(self) -> str:  # pragma: no cover - diagnostics only
        return f"DocRegistry(version={self.version!r}, groups={len(self.docs)})"

    def primary_url(self, key: str) -> str | None:
        entry = self.docs.get(key) or {}
        links = entry.get("links") or []
        if not links:
            return None
        url = links[0].get("url")
        return url if is_safe_url(url) else None

    def doc_for_source(self, source: str) -> dict[str, Any] | None:
        key = self.source_keys.get(source)
        return self.docs.get(key) if key else None

    def iter_links(self) -> Iterable[tuple[str, str, str, str]]:
        """Yield (group_key, group_name, link_label, url) for every link."""
        for group_key, entry in self.docs.items():
            name = str(entry.get("name", group_key))
            for link in entry.get("links", []) or []:
                yield group_key, name, str(link.get("label", "")), str(link.get("url", ""))

    def analyze(self, src: str | None, *, alt: str = "", source: str | None = None) -> dict[str, Any]:
        result = analyze(src, alt=alt, source=source)
        doc = self.doc_for_source(result["source"])
        if doc:
            key = self.source_keys[result["source"]]
            result["doc_ref"] = self.primary_url(key)
            result["doc_name"] = doc.get("name")
        return result


def load_registry(path: str | os.PathLike[str] | None = None) -> DocRegistry:
    """Load the generated registry JSON.

    Raises FileNotFoundError with an actionable message when the JSON has not
    been exported yet — run ``scripts/export_registry.mjs`` first.
    """
    target = Path(path) if path else DEFAULT_REGISTRY
    if not target.exists():
        raise FileNotFoundError(
            f"{target} not found — run `node scripts/export_registry.mjs` "
            "to export it from src/official-docs.js"
        )
    with target.open("r", encoding="utf-8") as handle:
        return DocRegistry(json.load(handle))
